README.md
# ula-vp-controller-plugin

[![Build Status](https://travis-ci.org/rabobank-blockchain/ula-vp-controller.svg?branch=master)](https://travis-ci.org/rabobank-blockchain/ula-vp-controller)
[![Test Coverage](https://api.codeclimate.com/v1/badges/c3583b99edad5c48168e/test_coverage)](https://codeclimate.com/github/rabobank-blockchain/ula-vp-controller/test_coverage)
[![Maintainability](https://api.codeclimate.com/v1/badges/c3583b99edad5c48168e/maintainability)](https://codeclimate.com/github/rabobank-blockchain/ula-vp-controller/maintainability)
[![JavaScript Style Guide](https://img.shields.io/badge/code_style-standard-brightgreen.svg)](https://standardjs.com)

This Holder [ULA](https://github.com/rabobank-blockchain/universal-ledger-agent) plugin responds to an incoming [ChallengeRequest](https://github.com/rabobank-blockchain/vp-toolkit-models/blob/master/src/model/challenge-request.ts) from the issuer and/or verifier.
This implementation uses secp256k1 by default. If you want to use a different cryptographic algorithm, then provide your own [crypt-util](https://github.com/rabobank-blockchain/crypt-util) instance.

## Installation

If you work with QR codes, please read the instructions in the [ula-process-eth-barcode](https://github.com/rabobank-blockchain/ula-process-eth-barcode) repository.
Also, in order to save credentials properly, we advise to install the [ula-vc-data-management](https://github.com/rabobank-blockchain/ula-vc-data-management) plugin as well.

In an existing project (with `package.json`), install the plugin by running the following commands:

```bash
npm install universal-ledger-agent --save
npm install crypt-util --save
npm install vp-toolkit --save
npm install ula-vp-controller-plugin --save
```

## Usage

This is an example of enabling this plugin in the ULA in a browser environment.

```typescript
import { BrowserHttpService, EventHandler } from "universal-ledger-agent";
import { VpController } from "ula-vp-controller";
import { LocalCryptUtils } from "crypt-util";
import {
  VerifiableCredentialGenerator,
  VerifiableCredentialSigner,
  VerifiablePresentationGenerator,
  VerifiablePresentationSigner
} from "vp-toolkit";

// Prepare plugin dependencies
const browserHttpService = new BrowserHttpService()
const cryptUtil = new LocalCryptUtils()
const vcSigner = new VerifiableCredentialSigner(cryptUtil)
const vcGenerator = new VerifiableCredentialGenerator(vcSigner)
const vpSigner = new VerifiablePresentationSigner(cryptUtil)
const vpGenerator = new VerifiablePresentationGenerator(vpSigner)
// Instantiate the plugin
const vpControllerPlugin = new VpController(cryptUtil, vcGenerator, vpGenerator, browserHttpService)

// Setup the ULA and inject the plugin
const ulaEventHandler = new EventHandler([vpControllerPlugin /*, other ULA plugins here */])
```

### Trigger

#### Using an existing ULA plugin
Add the [ula-process-eth-barcode](https://github.com/rabobank-blockchain/ula-process-eth-barcode) plugin to receive input from that plugin and you're set to go!

#### Manually
However, if you want to invoke this plugin manually, send a ULA message with this format:

```typescript
const msg = {
  type: 'process-challengerequest',
  msg: { /* IChallengeRequestParamsfields */
    toVerify: [{predicate: "{schema.org URL}", allowedIssuers: ["did:eth:allowedIssuer"]}, {predicate: "{schema.org URL}"}],
    toAttest: [{predicate: "{schema.org URL}"}, {predicate: "{schema.org URL}"}],
    postEndpoint: '{endpoint URL}', // The holder will post a VerifiablePresentation object here
    proof: { /* Proof object */ },
    nonce: "{uuid}"
  }
}

ulaEventHandler.processMsg(msg, (response: UlaResponse) => {
  // Handle callback
  // If statuscode is 1, update the UI
})
```
The `msg` property contains a [ChallengeRequest](https://github.com/rabobank-blockchain/vp-toolkit-models/blob/master/src/model/challenge-request.ts) object. Generate a ChallengeRequest object by using the [vp-toolkit](https://github.com/rabobank-blockchain/vp-toolkit) library.

Note: `toVerify` and `toAttest` fields are optional. You can use both at the same time or omit one of them. If you omit both fields, the plugin will complete without sending any visual feedback.

### Callbacks

When the plugin has finished, the callback function will be called twice.
The first time for updating the user interface and the second time for passing operation details (in this case, only a status code).

In case of a successful situation:
```typescript
// Dont show loading screen, show success screen
callback(new UlaResponse({ statusCode: 1, body: { loading: false, success: true, failure: false } }))
callback(new UlaResponse({ statusCode: 201, body: {} })) // Everything went OK
```

In case of an error:
```typescript
// Don't show loading screen, but show failure screen
callback(new UlaResponse({ statusCode: 1, body: { loading: false, success: false, failure: true } }))
callback(new UlaResponse({ statusCode: 204, body: {} })) // No information available
```

## Running tests

Besides unit testing with Mocha, the effectivity of all tests are also measured with the Stryker mutation testing framework.

```bash
npm run test
npm run stryker
```

We aim to achieve a coverage of 100%. Stryker and/or mocha test scores below 80% will fail the build.

## Contributing

Pull requests are welcome. For major changes, please open an issue first to discuss what you would like to change.

Please make sure to update tests as appropriate.

## License and disclaimer

[apache-2.0](https://choosealicense.com/licenses/apache-2.0/) with a [notice](NOTICE).

We discourage the use of this work in production environments as it is in active development and not mature enough.


HISTORY.md
# 0.2.5 / 24-07-2020

**Enhancements**
- Security patches updates from dependent packages

# 0.2.4 / 24-03-2020

**Enhancements**
- Add `sessionId` in the generated VP ([#25](https://github.com/rabobank-blockchain/ula-vp-controller/issues/25))
- Add predicate in self-attested VC ([#24](https://github.com/rabobank-blockchain/ula-vp-controller/issues/24))
- Updated dependencies
- Rollback Typescript to 3.4.5 to output proper `d.ts` files

# 0.2.3 / 10-03-2020

**Security fixes**
- Updated to `vp-toolkit@0.2.2` due to two Verifier vulnerabilities (not applicable here)

# 0.2.2 / 13-02-2020

**Bugfixes**
- When omitting the deprecated `msg.endpoint` field, the flow still exits prematurely ([#16](https://github.com/rabobank-blockchain/ula-vp-controller/issues/16))

# 0.2.1 / 08-01-2020

**Enhancements**
- Updated all dependencies
- Introduced [HISTORY.md](HISTORY.md)

# 0.2.0 / 27-12-2019

**BREAKING**
- Updated to `vp-toolkit-models@0.2.0`, introducing a new mandatory field in ChallengeRequest: `postEndpoint`. You need to add this field when you are sending a ULA message manually [(example here)](https://github.com/rabobank-blockchain/ula-vp-controller#manually)

**Enhancements**
- Fixed Handlebars vulnerability [CVE-2019-19919](https://github.com/advisories/GHSA-w457-6q6x-cgp9)

# 0.1.0 / 20-09-2019

*Initial release*


test/mock/cr-signer-mock.ts
/*
 *  Copyright 2020 Coöperatieve Rabobank U.A.
 *
 *  Licensed under the Apache License, Version 2.0 (the "License");
 *  you may not use this file except in compliance with the License.
 *  You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 */

import { CryptUtil } from 'crypt-util'
import { ChallengeRequestSigner } from 'vp-toolkit'

export class ChallengeRequestSignerMock extends ChallengeRequestSigner {

  constructor (private _cryptUtilImpl: CryptUtil) {
    super(_cryptUtilImpl)
  }

  get signatureType () {
    return 'WrongSignatureType'
  }
}


test/mock/vp-signer-mock.ts
/*
 *  Copyright 2020 Coöperatieve Rabobank U.A.
 *
 *  Licensed under the Apache License, Version 2.0 (the "License");
 *  you may not use this file except in compliance with the License.
 *  You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 */

import { CryptUtil } from 'crypt-util'
import { VerifiableCredentialSigner, VerifiablePresentationSigner } from 'vp-toolkit'

export class VerifiablePresentationSignerMock extends VerifiablePresentationSigner {

  constructor (private _cryptUtilImpl: CryptUtil, private _verifiableCredentialSignerImpl: VerifiableCredentialSigner) {
    super(_cryptUtilImpl, _verifiableCredentialSignerImpl)
  }

  get signatureType () {
    return 'WrongSignatureType'
  }
}


src/index.ts
/*
 *  Copyright 2020 Coöperatieve Rabobank U.A.
 *
 *  Licensed under the Apache License, Version 2.0 (the "License");
 *  you may not use this file except in compliance with the License.
 *  You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 */

export { VpController } from './vp-controller'

export { AddressHelper } from './service/address-helper'

export { VerifiableCredentialHelper } from './service/verifiable-credential-helper'


src/service/address-helper.ts
/*
 *  Copyright 2020 Coöperatieve Rabobank U.A.
 *
 *  Licensed under the Apache License, Version 2.0 (the "License");
 *  you may not use this file except in compliance with the License.
 *  You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 */

import { Address, IAddress } from 'ula-vc-data-management'
import { CryptUtil } from 'crypt-util'
import { EventHandler } from 'universal-ledger-agent'
import { VerifiableCredential } from 'vp-toolkit-models'

export class AddressHelper {

  constructor (private _cryptUtil: CryptUtil) {
  }

  /**
   * Generates a new address from the given accountId
   * and sends a save-address message to the ULA to
   * persist the info through another plugin.
   * (we advise to use ula-vc-data-management)
   *
   * @param {number} accountId
   * @param {string} predicate
   * @param  {EventHandler} eventHandler - used to
   * @return {Promise<string>} The generated public address
   */
  public async generateAndSaveAddressDetails (accountId: number, predicate: string, eventHandler: EventHandler): Promise<Address> {
    return new Promise(async (resolve) => {
      await eventHandler.processMsg({ type: 'get-new-key-id' },
        async (keyId: number) => {
          const pubAddress = this._cryptUtil.deriveAddress(accountId, keyId)
          const addressDetails: IAddress = {
            address: pubAddress,
            accountId: accountId,
            keyId: keyId,
            predicate: predicate
          }
          await eventHandler.processMsg({ type: 'save-address', address: addressDetails }, undefined)
          resolve(new Address(addressDetails))
        })
    })
  }

  /**
   * Get the address details for existing verifiable
   * credentials. The verifiable credential must be
   * in the format 'did:<anything>:<ethaddress>'
   *
   * @param {VerifiableCredential[]} credentials
   * @param {EventHandler} eventHandler
   * @return {Promise<void>}
   */
  public async findDidInfoForVCs (credentials: VerifiableCredential[], eventHandler: EventHandler): Promise<Address[]> {
    const promises: Promise<void>[] = []
    const didInfo: Address[] = []
    for (const credential of credentials) {
      const promise = new Promise<void>(async (resolve) => {
        const address = (credential.credentialSubject.id as string).split(':').pop()
        await eventHandler.processMsg({ type: 'get-address-details', publicAddress: address },
          async (address: Address) => {
            didInfo.push(address)
            resolve()
          })
      })
      promises.push(promise)
    }

    await Promise.all(promises)
    return didInfo
  }
}


src/service/verifiable-credential-helper.ts
/*
 *  Copyright 2020 Coöperatieve Rabobank U.A.
 *
 *  Licensed under the Apache License, Version 2.0 (the "License");
 *  you may not use this file except in compliance with the License.
 *  You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 */

import { ChallengeRequest, VerifiableCredential } from 'vp-toolkit-models'
import { EventHandler } from 'universal-ledger-agent'
import { Address } from 'ula-vc-data-management'
import { AddressHelper } from './address-helper'
import { VerifiableCredentialGenerator } from 'vp-toolkit'

export class VerifiableCredentialHelper {

  constructor (private _vcGenerator: VerifiableCredentialGenerator, private _addressHelper: AddressHelper) {
  }

  /**
   * Returns a collection of self-attested VC's to prove ownership over our DID's
   *
   * @param {ChallengeRequest} challengeRequest
   * @param {number} accountId provided by the wallet implementation
   * @param {EventHandler} eventHandler
   * @return {Promise<VerifiableCredential[]>} the self-attested VC's
   */
  public async generateSelfAttestedVCs (challengeRequest: ChallengeRequest, accountId: number, eventHandler: EventHandler):
    Promise<{ accountId: number, keyId: number, vc: VerifiableCredential }[]> {
    const verifiableCredentials: { accountId: number, keyId: number, vc: VerifiableCredential }[] = []
    for (const toAttest of challengeRequest.toAttest) {
      const addressDetails: Address = await this._addressHelper.generateAndSaveAddressDetails(accountId, toAttest.predicate, eventHandler)
      const did = 'did:eth:' + addressDetails.address
      const selfAttestedVc = this._vcGenerator.generateVerifiableCredential({
        type: ['VerifiableCredential', 'DidOwnership'],
        credentialSubject: {
          id: did,
          predicate: toAttest.predicate
        },
        '@context': [toAttest.predicate],
        issuanceDate: new Date(),
        issuer: did
      }, accountId, addressDetails.keyId)
      verifiableCredentials.push({ accountId: accountId, keyId: addressDetails.keyId, vc: selfAttestedVc })
    }

    return verifiableCredentials
  }

  /**
   * The verifier asks one or more VC's
   * using the toVerify field in the
   * ChallengeRequest. This method returns
   * a collection of VC's which match the
   * verifier's needs.
   *
   * @param {ChallengeRequest} challengeRequest
   * @param  {EventHandler} eventHandler
   * @return {Promise<VerifiableCredential[]>} the self-attested VC's
   */
  public async findVCsForChallengeRequest (challengeRequest: ChallengeRequest, eventHandler: EventHandler):
    Promise<{ matching: VerifiableCredential[], missing: { predicate: string, reason: string }[] }> {
    if (challengeRequest.toVerify.length === 0) {
      return {
        matching: [],
        missing: []
      }
    }

    const matchingVerifiableCredentials: { credential: VerifiableCredential, predicate: string }[] = []
    const failedCredentials: { credential: VerifiableCredential, predicate: string }[] = [] // These credentials do not pass the whitelist check
    const missingVCs: { predicate: string, reason: string }[] = []
    const regex = new RegExp('(' + challengeRequest.toVerify.map(value => value.predicate).join(')|(') + ')', 'g')
    await new Promise(async (resolve) => {
      // Get all credentials that match contain the predicate in their context arrays
      await eventHandler.processMsg(
        {
          type: 'get-vcs-by-context',
          contextRegex: regex
        },
        async (credentials: VerifiableCredential[]) => {
          // The challengeRequest.toVerify array contains predicate+whitelist

          // Step 1: Check which credentials do and don't pass the issuer whitelist check
          for (const toVerify of challengeRequest.toVerify) {
            let markAsMissing = true
            for (const credential of credentials) {
              const credentialSubjects = Object.keys(credential.credentialSubject)
              if (credentialSubjects.includes(toVerify.predicate)) {

                // Step 1.1: Incase the issuer whitelist is not present or it matches the whitelist,
                // mark the credential as 'keep this one'
                if ((toVerify.allowedIssuers === undefined
                  || toVerify.allowedIssuers.length === 0
                  || toVerify.allowedIssuers.includes(credential.issuer))) {
                  markAsMissing = false
                  matchingVerifiableCredentials.push({ credential: credential, predicate: toVerify.predicate })
                } else {
                  // Step 1.2: If the credential does not pass the whitelist check, mark it as 'remove this'.
                  // The credential can contain multiple subjects (claims), so one claim might pass the check
                  // while other claims might not. It is imperative that the credentials from step 2.1 are
                  // not removed. Only remove a credential if it doesn't pass for all claims completely, after
                  // this loop.
                  markAsMissing = false // Do not mark it as missing, let step 2 mark it differently
                  failedCredentials.push({ credential: credential, predicate: toVerify.predicate })
                }
              }
            }

            // Step 1.3: If there were no matching credentials found, add the predicate to the 'missing' array.
            // The user interface will show that there is no matching credential for this predicate.
            if (markAsMissing) {
              missingVCs.push({ predicate: toVerify.predicate, reason: 'missing' })
            }
          }

          // Step 2: Remove credentials if they don't match the whitelist check completely (for all credSubjects)
          for (const failedCred of failedCredentials) {
            if (!matchingVerifiableCredentials.map(x => x.predicate).includes(failedCred.predicate)
              && !this.containsMissingPredicate(missingVCs, failedCred.predicate)) {
              missingVCs.push({
                predicate: failedCred.predicate,
                reason: 'no-matching-issuer'
              })
            }
          }

          resolve()
        })
    })

    // Finally, return all matching credentials and missing predicates. Make sure not to send duplicated results.
    return {
      matching: this.getUniqueCredentials(matchingVerifiableCredentials),
      missing: missingVCs
    }
  }

  /**
   * Removes revoked credentials, saves issued
   * credentals and saves a Transaction object.
   *
   * @todo remove revoked credentials
   * @param {string} counterpartyId              The id of the counterparty
   * @param {string[]} verifiedVcs               Collection of VerifiableCredential nonces which were sent
   * @param {VerifiableCredential[]} credentials VP sent by the counterparty, containing attested VC's
   * @param {EventHandler} eventHandler          To send messages to the ULA data plugin
   */
  public async processTransaction (
    counterpartyId: string,
    verifiedVcs: string[],
    credentials: VerifiableCredential[],
    eventHandler: EventHandler) {
    await this.saveIssuedVCs(credentials, eventHandler)

    const transaction = {
      created: new Date(),
      counterpartyId: counterpartyId,
      state: 'success',
      issuedVcs: credentials.map(vc => vc.proof.nonce),
      verifiedVcs: verifiedVcs
    }
    await eventHandler.processMsg({ type: 'save-vc-transaction', transaction: transaction }, undefined)
  }

  /**
   * Save the Verifiable Credentials which
   * were sent by the issuer. A VC will only
   * be saved when the DID + predicate
   * matches with the address details in
   * storage.
   * This method does NOT verify any
   * signatures!
   *
   * @param {VerifiableCredential[]} credentials The Verifiable Presentation from the issuer
   * @param {EventHandler} eventHandler
   */
  public async saveIssuedVCs (credentials: VerifiableCredential[], eventHandler: EventHandler) {
    const promises: Promise<void>[] = []
    const vcsToSave: VerifiableCredential[] = []
    for (const vc of credentials) {
      // Todo: Implement universal resolver to resolve the DID
      // Right now, we assume that the given DID is in the following format - did:xx:publicAddress
      const pubAddress = (vc.credentialSubject.id as string).split(':').pop()
      promises.push(new Promise(async (resolve) => {
        await eventHandler.processMsg(
          { type: 'get-address-details', publicAddress: pubAddress },
          async (address: Address) => {
            if (vc.context && vc.context.includes(address.predicate)) {
              vcsToSave.push(vc)
            }
            resolve()
          })
      }))
    }

    await Promise.all(promises)
    await eventHandler.processMsg({ type: 'save-vcs', verifiableCredentials: vcsToSave }, undefined)
  }

  private getUniqueCredentials (array: { credential: VerifiableCredential; predicate: string }[]) {
    return [...new Set(array.map(x => x.credential))]
  }

  private containsMissingPredicate (array: { predicate: string, reason: string }[], predicate: string) {
    return array.filter(x => x.predicate === predicate).length > 0
  }
}


src/vp-controller.ts
/*
 *  Copyright 2020 Coöperatieve Rabobank U.A.
 *
 *  Licensed under the Apache License, Version 2.0 (the "License");
 *  you may not use this file except in compliance with the License.
 *  You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 */

import { EventHandler, HttpService, Message, Plugin, UlaResponse } from 'universal-ledger-agent'
import {
  ChallengeRequest,
  IVerifiablePresentationParams,
  VerifiableCredential,
  VerifiablePresentation
} from 'vp-toolkit-models'
import { ChallengeRequestSigner, VerifiablePresentationGenerator, VerifiablePresentationSigner } from 'vp-toolkit'
import { VerifiableCredentialHelper } from './service/verifiable-credential-helper'
import { AddressHelper } from './service/address-helper'

/**
 * The VP Controller ULA plugin
 * ensures a correct issue/verify flow
 * by delegating various activities to
 * the correct dependencies.
 */
export class VpController implements Plugin {
  private _eventHandler?: EventHandler

  /**
   * Provide the generators you want to use
   * in order to generate VerifiableCredentials
   * and VerifiablePresentations.
   *
   * Multiple signers for one class can be provided,
   * so you can verify objects which were signed with
   * different algorithms. If the VerifiablePresentation
   * from the issuer does not contain any proofs, the
   * first given VerifiablePresentationSigner will be
   * used. Create your own signer by overriding the
   * existing signer class.
   *
   * The account ID is the 'wallet' or 'profile'
   * identifier the current user is utilizing.
   * If your wallet implementation does not provide
   * multiple wallets/profiles, then you can
   * provide 0 as accountId value.
   *
   * @param {VerifiablePresentationGenerator} _vpGenerator
   * @param {VerifiablePresentationSigner[]} _vpSigners
   * @param {ChallengeRequestSigner[]} _challengeRequestSigners
   * @param {HttpService} _httpService
   * @param {VerifiableCredentialHelper} _vcHelper
   * @param {AddressHelper} _addressHelper
   * @param {number} _accountId
   */
  public constructor (
    private _vpGenerator: VerifiablePresentationGenerator,
    private _vpSigners: VerifiablePresentationSigner[],
    private _challengeRequestSigners: ChallengeRequestSigner[],
    private _httpService: HttpService,
    private _vcHelper: VerifiableCredentialHelper,
    private _addressHelper: AddressHelper,
    private _accountId: number
  ) {
  }

  /**
   * The name of the plugin
   * @return {string}
   */
  get name () {
    return 'VpController'
  }

  /**
   * The current wallet/profile ID
   * @return {number}
   */
  get accountId () {
    return this._accountId
  }

  /**
   * When the user switches to a different
   * wallet/profile, update the accountId
   * to make sure the correct keys are
   * generated and used.
   */
  set accountId (id: number) {
    this._accountId = id
  }

  /**
   * Receive the eventHandler so we can put messages
   * back on the ULA again
   * @param {EventHandler} eventHandler
   */
  public initialize (eventHandler: EventHandler): void {
    this._eventHandler = eventHandler
  }

  /**
   * Handle incoming messages
   * @param {Message} message
   * @param callback
   * @return {Promise<string>}
   */
  public async handleEvent (message: Message, callback: any): Promise<string> {
    if (message.properties.type.match('accept-consent')) {
      return this.handleConsent(message, callback)
    }

    if (message.properties.type !== 'process-challengerequest') {
      return 'ignored' // This message is not intended for us
    }

    if (!message.properties.msg) {
      return 'ignored' // The message type is correct, but endpoint or msg is missing
    }

    if (!this._eventHandler) {
      this.triggerFailure(callback)
      throw new Error('Plugin not initialized. Did you forget to call initialize() ?')
    }

    try {
      const challengeRequest = new ChallengeRequest(message.properties.msg)
      // Check if we expect a response containing a VP with issued VC's from the issuer (otherwise it is a verifier)
      const matchingCrSigner = this._challengeRequestSigners.find((crSigner) => crSigner.signatureType === challengeRequest.proof.type)
      const isValidChallengeRequest = matchingCrSigner ? matchingCrSigner.verifyChallengeRequest(challengeRequest) : false

      if (!isValidChallengeRequest) {
        this.triggerFailure(callback)
        return 'error-cr'
      }

      // toAttest process
      // Receive the DidInfo to create a new proof using the same DID keys
      const selfAttestedVCsAndDidInfo = await this._vcHelper.generateSelfAttestedVCs(challengeRequest, this._accountId, this._eventHandler)
      const selfAttestedVCs = selfAttestedVCsAndDidInfo.map((vcd: any) => vcd.vc)

      // toVerify process
      const vcSearchResult = await this._vcHelper.findVCsForChallengeRequest(challengeRequest, this._eventHandler)
      const existingVcAddresses = await this._addressHelper.findDidInfoForVCs(vcSearchResult.matching, this._eventHandler)

      // Transform all DID info so generateVerifiablePresentation can digest it
      const selfAttestedDidInfo = selfAttestedVCsAndDidInfo.map(info => {
        return { accountId: info.accountId, keyId: info.keyId }
      })
      const existingVcDidInfo = existingVcAddresses.map(addr => {
        return { accountId: addr.accountId, keyId: addr.keyId }
      })

      // Prepare the response, but make the VP undefined if there are no credentials found
      const allFoundCredentials = selfAttestedVCs.concat(vcSearchResult.matching)
      const selfAttestedVP = allFoundCredentials.length > 0 ?
        this._vpGenerator.generateVerifiablePresentation(
          {
            type: ['VerifiablePresentation', 'ChallengeResponse'],
            verifiableCredential: allFoundCredentials,
            sessionId: challengeRequest.correspondenceId
          } as IVerifiablePresentationParams,
          selfAttestedDidInfo.concat(existingVcDidInfo),
          challengeRequest.correspondenceId
        )
        : undefined

      // Ask for consent
      const nextMessage = new UlaResponse(
        {
          statusCode: 200,
          body: {
            confirmAttestations: vcSearchResult.matching.map((vc) => {
              for (const cs of Object.keys(vc.credentialSubject)) {
                if (cs !== 'id') {
                  return {
                    key: cs.split('/').pop(),
                    value: vc.credentialSubject[cs],
                    attestor: vc.additionalFields['issuerName']
                  }
                }
              }
            }),
            missingAttestations: vcSearchResult.missing,
            filledTemplate: {
              challengeRequest: challengeRequest,
              verifiablePresentation: selfAttestedVP
            },
            url: challengeRequest.postEndpoint,
            type: 'accept-consent'
          }
        }
      ) // Todo: Redesign this message structure

      // If the counterparty requests data (toVerify), show the consent screen
      if (challengeRequest.toVerify.length > 0) {
        callback(new UlaResponse({ statusCode: 1, body: { loading: false, success: false, failure: false } }))
        callback(nextMessage)
      } else {
        nextMessage.body.payload = nextMessage.body.filledTemplate
        return this.handleConsent(new Message(nextMessage.body), callback)
      }

    } catch (error) {
      this.triggerFailure(callback)
      return 'error'
    }

    return 'success'
  }

  private async handleConsent (message: Message, callback: any): Promise<string> {
    // Send challengeresponse (VP) and process the response from the endpoint
    const challengeRequest = message.properties.payload.challengeRequest as ChallengeRequest
    const selfAttestedVP = message.properties.payload.verifiablePresentation as VerifiablePresentation
    const response = await this._httpService.postRequest(challengeRequest.postEndpoint, selfAttestedVP)
    let issuedCredentials: VerifiableCredential[] = []

    // The endpoint can either be an issuer sending back a VP - or a verifier sending back an empty response
    if (challengeRequest.toAttest.length > 0) {
      const vp = new VerifiablePresentation(response as IVerifiablePresentationParams)
      issuedCredentials = vp.verifiableCredential
      const matchingVpSigner = this._vpSigners.find((vpSigner) => vp.proof.length > 0 && vpSigner.signatureType === vp.proof[0].type)
      const vpIsValidVp = matchingVpSigner
        ? matchingVpSigner.verifyVerifiablePresentation(vp, true)
        : this._vpSigners[0].verifyVerifiablePresentation(vp, true)

      if (!vpIsValidVp) {
        this.triggerFailure(callback)
        return 'error-vp'
      }
    }

    // Save the VC's coming from the issuer
    await this._vcHelper.processTransaction(
      challengeRequest.proof.verificationMethod,
      selfAttestedVP.verifiableCredential.filter(vc => (!vc.type.includes('DidOwnership'))).map(vc => vc.proof.nonce),
      issuedCredentials,
      // @ts-ignore
      this._eventHandler
    )

    callback(new UlaResponse({ statusCode: 1, body: { loading: false, success: true, failure: false } }))
    callback(new UlaResponse({ statusCode: 201, body: {} }))

    return 'success'
  }

  private triggerFailure (callback: any) {
    if (callback) {
      callback(new UlaResponse({ statusCode: 1, body: { loading: false, success: false, failure: true } }))
      callback(new UlaResponse({ statusCode: 204, body: {} }))
    }
  }
}
